### ISSUE TITLE
[Linux] Exploding vertices: Ni No Kuni

### ISSUE THREAD

**Issue by hcorion** [RPCS3/rpcs3#2677]
Seen working correctly on windows in this video: https://www.youtube.com/watch?v=KyXP11pVIjQ
~~If you use ASMJIT or fast SPU the character's are invisible, but using precise SPU is supposed to fix it, but it completely screws up for me.~~
**EDIT:**
With PR #2683  the issue now occurs with all SPU options.

Specs:
i7-4770 (non-k)
GTX 1070
Arch Linux 64-bit
KDE Plasma
Nvidia beta driver version: 381.09 (same problem with stable driver)
RPCS3 v0.0.2-5124-3ce7947d Alpha

Log https://gist.githubusercontent.com/anonymous/744db98999bb6a88d7b5ff7014ed544e/raw/6c4ba19a347eb91a3197fc08d42aa587584149f5/RPCS3.log
Stopped emulator at point in second screenshot.

Screenshots:
![](https://cdn.discordapp.com/attachments/272035812277878785/301069061746458624/unknown.png)

![](https://cdn.discordapp.com/attachments/272035812277878785/301068774331777024/unknown.png)

Custom cmake flags:
```shell
cmake  -DCMAKE_CXX_FLAGS='-march=native -mtune=native -flto -fuse-ld=gold' -DCMAKE_EXE_LINKER_FLAGS='-ldl -lyaml-cpp -march=native -mtune=native -flto -fuse-ld=gold' -DCMAKE_SHARED_LINKER_FLAGS='-march=native -mtune=native -flto -fuse-ld=gold' \
 -DCMAKE_C_FLAGS='-march=native -mtune=native -flto -fuse-ld=gold' \
 ../../source/rpcs3 -G "Unix Makefiles" \
 -DUSE_SYSTEM_LIBPNG=ON -DUSE_SYSTEM_FFMPEG=ON \
 -DCMAKE_BUILD_TYPE=Release -DCMAKE_SKIP_RPATH='ON'
```
(basically just link time optimization, use native cpu features, use gold linker and use system ffmpeg and libpng)

**Comment by jjsat** [comment]
I wonder if this is only caused by different optimization of floating point math between GCC and MS VC. For MS VC the "precise" optimization preset is used and that should not do standard-violating optimizations. GCC seems to not be fully compliant by default (https://gcc.gnu.org/wiki/FloatingPointMath). Maybe the result is better with "-frounding-math -fsignaling-nans" options.

**Comment by hcorion** [comment]
@jjsat I'll try those compile flags when I have some time.

**Comment by hcorion** [comment]
@jjsat 
After cleaning the cmake cache and re-compiling with -frounding-math and -fsignaling-nans it doesn't make a difference.

**Comment by hcorion** [comment]
With PR #2683 the issue now occurs with fast SPU and ASMJIT SPU.

**Comment by hcorion** [comment]
Ok, so I compiled with clang 3.9.1, and it works perfectly, so it seems to be a GCC-only bug.

**Comment by paulsapps** [comment]
try with ubsan enabled

**Comment by hcorion** [comment]
@paulsapps Yep, compiling with `-fsanitize=undefined` fixes it for GCC 6.3.1, however, performance is much diminished.

**Comment by paulsapps** [comment]
It shouldn't fix anything.. it should hopefully find some UB which once resolved "fixes" it :P

**Comment by kirbyfan64** [comment]
Smells like an optimization issue to me...why else would it disappear when compiled with the sanitizer?

Maybe Valgrind would be a bit more useful in this case?

**Comment by paulsapps** [comment]
Either its an optimization issue caused by compiler bug or probably some UB/IB. Unless its race condition fixed by ubsan slowdown?

**Comment by hcorion** [comment]
RPCS3 has been known to find a lot of corner cases in compilers, another interesting case is that the LLVM recompiler doesn't work with ubsan.

**Comment by hcorion** [comment]
Midnight Club LA is also affected by this bug. Works fine in clang, exploding vertices when compiled with GCC (7.2.0).

**Comment by HenningJW** [comment]
It's probably possible to find the translation unit that causes that by bisecting with selective activation of "#pragma GCC ..."?

**Comment by hcorion** [comment]
This seems to have fixed itself, I'm assuming it's gcc 7.3.0 that fixed it.
